keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,368 @@
+**Motivation:**  less downtime, optimal use of resources<br/>
+**How Resource classes can solve this:** Explained above<br/>
+**Can this be solved without resource classes:** No
+
+### As a developer:
+- I want the ability to be able to request devices which have specific capabilities. Eg: GPUs that are Volta or newer.<br/>
+  **Motivation:** I want minimum guaranteed compute performance<br/>
+  **Can this be solved without resource classes:**<br/>
+  - Yes, using node labels and NodeLabelSelectors.
+    Problem: Same problem of lack of access control on using labelselectors at user level as with the use of tolerations.
+  - OR, Instead of using resource class, provide flexibility to query resource properties directly in pod container resource requests.
+    Problem: In a large cluster, computing operators like “greater than”, “less than” at pod creation can be a very slow operation and is not scalable.
+
+  **How Resource classes can solve this:**
+The Kubernetes scheduler is the central place to map container resource requests expressed through ResourceClass names to the underlying qualified physical resources, which automatically supports metadata aware resource scheduling.
+
+- As a data scientist, I want my workloads to use advanced compute resources available in the running clusters without understanding the underlying hardware configuration details. I want the same workload to run on either on-prem Kubernetes clusters or on cloud, without changing its pod spec. When a new hardware driver comes out, I hope all the required resource configurations are handled properly by my cluster operators and things will just continue to work for any of my existing workloads.<br/>
+**Motivation:** Separation of concerns between cluster operators and cluster users.<br/>
+**Can this be solved without resource classes:**<br/>
+Without the additional abstraction layer, consuming the non-standard, metadata-rich compute resources would be fragmented. More likely, we would see cluster providers implement their own solutions to address their user pains, and it would be hard to provide a consistent user experience for consuming extended resources in the future.
+
+### As a vendor:
+- I want an easy and extensible mechanism to export my resource to Kubernetes. I want to be able to roll out new hardware features to the users who require those features without breaking users who are using old versions of hardware.<br/>
+**Motivation:** enables more compute resources and their advanced features on Kubernetes<br/>
+**Can this be solved without resource classes:**<br/>
+Yes, Using node labels and NodeLabelSelectors.<br/>
```

Comment:
Again seems pretty vague. Is this a real need today? The example mentioned below also doesn't seem realistic. 
Are there use cases outside of GPUs?

Reply:
This user story is mostly motivated by some past discussions on device plugin features requests, e.g., as @kad  mentioned in https://github.com/kubernetes/kubernetes/issues/59109#issuecomment-365750614
I like the explicit API model that once ResourceClass is in place, Kubelet can pass ResourceClass name to a device plugin, and the device plugin can map that ResourceClass name to the special underlying resource metadata requirements.

Reply:
Another example could be different types of cpu cores i,e isolated cores, numa affined cores, hyperthreaded cores etc.

Reply:
For https://github.com/kubernetes/kubernetes/issues/59109#issuecomment-365750614 wouldn't Pod annotations suffice?

> Another example could be different types of cpu cores i,e isolated cores, numa affined cores, hyperthreaded cores etc.

I don't think we are decided yet on what level of CPU specifics we want to expose to users.

Reply:
After cpu manager's static policy, IMO, supporting features like isolated cores is a natural progression.
/cc @jeremyeder

Reply:
i think this feature is orthogonal to cpus and confuses the discussion.